dataflow_test.py

```diff
@@ -357,6 +357,19 @@ def test_branch():
     assert e1 == e2
 
 
+def test_branch_closes_sideways():
+    the_source = range(10)
+    branch_result = []; the_branch_sink = df.sink(branch_result.append)
+    main_result   = []; the_main_sink   = df.sink(  main_result.append)
+
+    df.push(source = the_source,
+            pipe   = df.pipe(df.branch(the_branch_sink),
+                             the_main_sink))
+
+    with raises(StopIteration):
+        the_branch_sink.send(99)
```

Comment:
Hmm ... is this enough? The test name implies (at least to me) that closing the branch should close other things which would otherwise remain open. I don't see the body of the test addressing that kind of stuff at all.

I'll need to think about this more carefully. Don't have time now.

[Aside: before I forget ... I suspect that the `close_all` option in `slice` is not tested adequately]

Reply:
Well, this test was particularly hard to develop because I am not yet familiar with the whole coroutine thing. The idea was that originally we discovered that when doing a branch the sideways pipeline was not closed at the end of a push, but I don't remember how or what implication it had. So I started playing with it a bit and I realized that not closing that pipeline means that you could still send data through it, so that's what I attempted here.

I think in order to write a good test for this I will need your help, so we can wait until you find the time for it.

> Aside: before I forget ... I suspect that the close_all option in slice is not tested adequately

Yes. I've found a comment in the code saying so, I will add it to the list.

Reply:
Whilst you're waiting for me, and only if you feel like it, try

1. Checking the *other* branch. At the moment you're checking the side-branch, but not the main one.

Did your reverted test-failure-provocation make anything fail?

One likely consequence of some branch not being closed would be any future associated with the sink at the end of the branch, not having a value available.